core/trino-main/src/main/java/io/trino/execution/scheduler/PipelinedStageExecution.java

```diff
@@ -380,15 +370,49 @@ private synchronized void updateTaskStatus(TaskStatus taskStatus)
             if (taskState == TaskState.RUNNING) {
                 stateMachine.transitionToRunning();
             }
-            if (isFlushing()) {
+            if (isStageFlushing()) {
                 stateMachine.transitionToFlushing();
             }
-            if (finishedTasks.containsAll(allTasks)) {
+            if (isStageFinished()) {
                 stateMachine.transitionToFinished();
             }
         }
     }
 
+    private synchronized boolean isStageFlushing()
+    {
+        // to transition to flushing, there must be at least one flushing task, and all others must be flushing or finished.
+        return !flushingTasks.isEmpty()
+                && allTasks.stream().allMatch(taskId -> finishedTasks.contains(taskId) || flushingTasks.contains(taskId));
+    }
+
+    private synchronized boolean isStageFinished()
+    {
+        return finishedTasks.containsAll(allTasks);
+    }
+
+    private boolean addFlushingTask(TaskId taskId)
+    {
+        if (!flushingTasks.contains(taskId)) {
+            synchronized (this) {
+                return flushingTasks.add(taskId);
+            }
+        }
+        return false;
+    }
+
+    private boolean addFinishedTask(TaskId taskId)
+    {
+        boolean changed = false;
+        if (!finishedTasks.contains(taskId) || flushingTasks.contains(taskId)) {
```

Comment:
just `if (!finishedTasks.contains(taskId))` is sufficient

Reply:
No, it is not (at least on my eye). There is a race condition here - I mean this method handles changing of task statuses. On my eye it is possible that we could observer a such order about specific `taskId`.
`FLUSHING`, `FINISHED`, `FLUSHING`, `FINISHED`. It could happen because of network latencies / request retrials. Then, we need a possibility to move the task from `FLUSHING` to `FINISHED` again. 
It is not something new - I mean there is exactly same issue in the fully synchronized version.

Reply:
But you don't test if the `FINISHED` task is set again to `FLUSHING`.

Reply:
Yes, but the previous version (before the change) does not it as well.

Reply:
task shouldn't move back to `flushing` state. Therefore `finishedTasks.contains(taskId) && flushingTasks.contains(taskId)` is not a valid state

Reply:
Yes, it shoud not. But, as we were discussing it is possbile that we process task statuses in the out of order manner.